pyneuroml/swc/LoadSWC.py

```diff
@@ -0,0 +1,292 @@
+
+            for child in children:
+                fraction_along = child.fraction_along
+                distance_to_proximal = parent_length * fraction_along
+                graph.add_edge(parent_id, child.id, weight=distance_to_proximal)
+
+        self.graph = graph
+        return graph
+
+    def get_parent(self, node_id):
+        """Get the parent of a given node."""
+        node = self.nodes.get(node_id)
+        if node is None:
+            raise ValueError(f"Node {node_id} not found")
+        return self.nodes.get(node.parent_id) if node.parent_id != -1 else None
+
+    def get_descendants(self, node_id):
+        """Get all descendants of a given node."""
+        node = self.nodes.get(node_id)
+        if node is None:
+            raise ValueError(f"Node {node_id} not found")
+
+        descendants = []
+        queue = node.children.copy()
+        while queue:
+            child_id = queue.pop(0)
+            child = self.nodes[child_id]
+            descendants.append(child)
+            queue.extend(child.children)
+        return descendants
+
+    def get_nodes_with_multiple_children(self, type_id=None):
+        """Get all nodes that have more than one child, optionally filtering by type."""
+        nodes = []
+        for node in self.nodes.values():
+            if len(node.children) > 1:
+                if type_id is None or node.type == type_id:
+                    nodes.append(node)
+        return nodes
+
+    def get_nodes_by_type(self, type_id):
+        """Get all nodes of a specific type."""
+        return [node for node in self.nodes.values() if node.type == type_id]
+
+    def get_branch_points(self, *types):
+        """Get all branch points (nodes with multiple children) of the given types."""
+        nodes = []
+        for type_id in types:
+            nodes.extend(self.get_nodes_with_multiple_children(type_id))
+        return nodes
+
+    def has_soma_node(self):
+        """
+        Check if the graph contains at least one soma node (type 1).
+        """
+        return any(node.type == SWCNode.SOMA for node in self.nodes.values())
+
+
+def parse_header(line):
+    for field in SWCGraph.HEADER_FIELDS:
+        match = re.match(rf"{field}\s+(.+)", line, re.IGNORECASE)
+        if match:
+            return field, match.group(1).strip()
+    return None, None
+
+
+def load_swc(filename):
+    tree = SWCGraph()
+    try:
+        with open(filename, "r") as file:
+            for line in file:
+                line = line.strip()
+                if not line:
+                    continue
+                if line.startswith("#"):
+                    key, value = parse_header(line[1:].strip())
+                    if key:
+                        tree.add_metadata(key, value)
+                    continue
+
+                parts = line.split()
+                if len(parts) != 7:
+                    print(f"Warning: Skipping invalid line: {line}")
```

Comment:
I'd make this a logger.error or logger.warning perhaps, and we need to decide if this should stop the parsing of the file.

Reply:
this still needs to be updated @AdityaBITMESRA